app/src/main/java/com/example/sw0b_001/OTPVerificationActivity.kt

```diff
@@ -86,6 +117,28 @@ class OTPVerificationActivity : AppCompactActivityCustomized() {
         vault = Vault(applicationContext)
     }
 
+    private fun startCountdownTimer(nextAttemptTimestamp: Long, onTick: (Long) -> Unit, onFinish: () -> Unit) {
+        val currentTimeMillis = System.currentTimeMillis()
+        val timeDiffInMillis = nextAttemptTimestamp * 1000 - currentTimeMillis
+
+        if (timeDiffInMillis > 0) {
+            object : CountDownTimer(timeDiffInMillis, 1000) {
+                override fun onTick(millisUntilFinished: Long) {
+                    val secondsRemaining = millisUntilFinished / 1000
+                    onTick(secondsRemaining)
+                }
+
+                override fun onFinish() {
+                    onFinish()
+                    finish()
+                    Toast.makeText(this@OTPVerificationActivity, R.string.sms_code_expired, Toast.LENGTH_LONG).show()
```

Comment:
When the timer reaches zero, it has not expired - it's a timer till when the user can request a new one. So do not finish the activity, they may still receive the code. Let them choose the resend button (or whatever UI) before you can finish the activity.

Put all the code for the onFinish() inside the callback